pkg/services/k8s/wrappers/dashboards.go

```diff
@@ -0,0 +1,187 @@
+		return nil, fmt.Errorf("no dashboard data")
+	}
+
+	namespace := util.OrgIdToNamespace(cmd.OrgID)
+	dashboardResource, err := s.clientset.GetClientset().GetResourceClient(schema.GroupVersionKind{
+		Group:   "dashboard.kinds.grafana.com",
+		Version: "v0-alpha",
+		Kind:    "Dashboard",
+	}, namespace)
+	if err != nil {
+		return nil, fmt.Errorf("ProvideServiceWrapper failed to get dashboard resource client: %w", err)
+	}
+
+	res := genericDashboardResource{
+		Metadata: kinds.GrafanaResourceMetadata{
+			Annotations: make(map[string]string),
+		},
+	}
+	res.Metadata.SetCommitMessage(cmd.Message)
+
+	// Set the resource folder
+	if cmd.FolderUID != "" {
+		res.Metadata.SetFolder(cmd.FolderUID)
+	} else if cmd.FolderID > 0 {
+		f, err := s.folders.GetFolderByID(ctx, cmd.OrgID, cmd.FolderID)
+		if err != nil {
+			return nil, err
+		}
+		res.Metadata.SetFolder(f.UID)
+	}
+
+	if provisioning != nil {
+		res.Metadata.SetOriginInfo(&kinds.ResourceOriginInfo{
+			Name:      provisioning.Name,
+			Path:      provisioning.ExternalID,
+			Key:       provisioning.CheckSum,
+			Timestamp: util.Pointer(time.UnixMilli(provisioning.Updated)),
+		})
+	} else if cmd.PluginID != "" {
+		res.Metadata.SetOriginInfo(&kinds.ResourceOriginInfo{
+			Name: "plugin",
+			Key:  cmd.PluginID,
+		})
+	}
+
+	res.Metadata.Namespace = util.OrgIdToNamespace(cmd.OrgID)
+
+	dto := cmd.GetDashboardModel()
+	if dto.UID == "" {
+		res.Metadata.GenerateName = "auto" // will create a new UID
+	} else {
+		res.Metadata.Name = dto.UID
+		res.Metadata.GenerateName = ""
+	}
+
+	if cmd.IsFolder {
+		res.APIVersion = "folders.kinds.grafana.com/v0-alpha"
+		res.Kind = "Folder"
+		res.Spec = cmd.Dashboard
+
+		if provisioning == nil {
+			return s.DashboardSQLStore.SaveDashboard(ctx, cmd)
+		}
+		return s.DashboardSQLStore.SaveProvisionedDashboard(ctx, cmd, provisioning)
+	}
+
+	// Dashboard versions
+	res.APIVersion = "dashboard.kinds.grafana.com/v0-alpha"
+	res.Kind = "Dashbaord"
+	res.Spec = cmd.Dashboard
+
+	rv := ""
+	r, err := dashboardResource.Get(ctx, dto.UID, metav1.GetOptions{})
+	if err != nil || r == nil {
+		fmt.Printf("did not find: %s\n", dto.UID)
+	} else {
+		if !cmd.Overwrite {
+			fmt.Printf("TODO... verify SQL version: %s\n", r.GetResourceVersion())
+		}
+		rv = r.GetResourceVersion()
+		res.Metadata.ResourceVersion = rv
+	}
+
+	out, err := runtime.DefaultUnstructuredConverter.ToUnstructured(&res)
+	if err != nil {
+		return nil, err
+	}
+	uObj := &unstructured.Unstructured{Object: out}
+	if rv == "" {
+		s.log.Debug("k8s action: create")
+		uObj, err = dashboardResource.Create(ctx, uObj, metav1.CreateOptions{})
+	} else {
+		s.log.Debug("k8s action: update")
+		uObj, err = dashboardResource.Update(ctx, uObj, metav1.UpdateOptions{})
+	}
+	if err != nil {
+		return nil, err
+	}
+	js, _ := json.Marshal(uObj)
+	err = json.Unmarshal(js, &res)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO! depend on the write command above to make synchronous call to SQL storage
```

Comment:
Is this still a TODO? I am having trouble understanding how the Create/Update call to apiserver would effect into writing to to-be-legacy SQL storage? Isn't apiserver not aware of the legacy SQL storage?

Reply:
Yes, still a TODO -- because the write within storage is still a noop, we are explicitly calling here.

> So it seems we are tending towards making the apiserver code in Grafana aware of to-be-legacy SQL storage so it can manage that store in addition to it's own store?

As a migration path, i think this gives us the most reasonable way to make sure current SQL is the source of truth.